[자동차 경주] 워터 제출합니다.

osandra opened the pull request:
안녕하세요! 
PR 리뷰는 처음이네요 ㅎㅎ 
변경하면 좋을 부분, 제안하는 부분, 궁금한 부분 모든 물어봐 주세요. 고쳐나가겠습니다 ㅎㅎ

- kotest를 사용하여 테스트 코드를 작성했습니다.(관련 라이브러리 2개를 build.gradle에 추가했습니다)
- 클래스 구성
    - Car: 자동차 객체
    - InputDto: 입력받은 자동차 이름 리스트와 횟수를 전달
    - GameConsoleInput: 유저에게 입력값을 입력받아 InputDto 형태로 반환
    - RandomGenerator: 0과 9사이 랜덤 숫자 생성하여 반환
    - GameApplication: 입력받은 횟수만큼 게임을 진행하고 각 단계 및 결과를 출력

osandra committed:
feat: 콘솔로 입력받는 부분 코드 추가
자동차 이름, 이동 횟수 콘솔 입력 부분 코드 및 build.gradle(kotest) 추가

osandra committed:
test: 콘솔로 입력받는 부분 테스트 코드 추가
자동차 이름 쉼표 구분, 길이 제한 및 이동 횟수 값 숫자 확인 테스트 코드 추가

osandra committed:
style: 콘솔 입력 부분 패키지 변경 (racingcar -> racingcar/ui)

osandra committed:
feat: Car 클래스 및 RandomGenerator 클래스 코드 추가

osandra committed:
test: Car 클래스 및 RandomGenerator 테스트 코드 추가

osandra committed:
feat: Game 진행 관련 코드 추가
횟수만큼 레이스 진행과 각 레이스 진행 후 자동차 출력 및 우승자를 출력하는 부분 코드 추가

osandra committed:
test: Game 진행 관련 테스트 코드 추가

osandra committed:
fix: 적절한 안내 메시지 출력 및 Car 클래스의 프로퍼티 setter 설정 추가

osandra committed:
refactor: 변수명 변경

perenok commented on build.gradle:
```diff
@@ -23,4 +24,4 @@ compileTestKotlin {
 test {
     useJUnitPlatform()
-}
+}
```
여기부터 여러 파일들에 아래 공백이 없는 것이 보이는데요ㅎㅎ
파일 마지막에 빈 줄을 넣는 이유에 대해 한번 찾아보시면 좋을 거 같아요!
그리고 빈 줄이 없는 채로 push하게 되면 깃허브에서 No newline at end of file이라고 경고가 뜹니다!

perenok commented on src/test/kotlin/water/racingcar/util/RandomGeneratorTest.kt:
```diff
@@ -0,0 +1,15 @@
+    given("RandomGeneratorTest 클래스에서"){
+        `when`("랜덤 번호를 생성하면"){
+            then("랜덤 숫자는 0이상 9이하여야 한다."){
+                randomGenerator.generateRandomNumber() shouldBeInRange 0..9
```
generateRandomNumber()를 통해 생성된 하나의 넘버가 테스트를 통과한다고 매번 랜덤하게 생성되는 숫자가 다 통과할 수 있을지 의문이 들어요.
물론 간단한 코드라 저희 모두 0~9 안에 숫자가 나올거라는 것을 예상할 수 있지만ㅋㅋㅋ 100% 확실하게 제어할 수 없는 이러한 코드를 테스트하는 거에 대해서 저는 반대의견인데 워터의 생각은 어떤가요?

osandra replied:
100% 제어할 수 없는 코드를 테스트하는 것에 대해 생각해보지 못했네요! 확인하고자 하는 값이 나오지 않을 수도 있으니 제어할 수 없는 부분은 테스트 하지 않고 CarTest 에서 움직이는 부분에 대한 테스트코드만 작성하는 방향이 좋겠네요! 좋은 의견 감사해요 ㅋㅋ

perenok commented on src/main/kotlin/water/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,23 @@
+        var moved = movedValue
+
+        output.append("$name : ")
+        while (moved > 0) {
+            output.append("-")
+            moved--
+        }
```
```suggestion
        repeat(movedValue) {
            output.append("-")
        }
```
이렇게 바꾸면 moved를 따로 선언 안 해줘도 원하는 결과를 얻을 수 있을 거 같아요ㅎㅎ

osandra replied:
훨씬 좋은 방법이 있었군요!! 수정했습니다 ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/dto/InputDto.kt:
```diff
@@ -0,0 +1,3 @@
+package water.racingcar.dto
+
+data class InputDto(val carNames: List<String>, val times: Int )
```
Dto 네이밍에 좀 더 의미가 포함되면 좋을 것 같아요ㅎㅎ
현재 InputDto 클래스만 봤을 땐 명확히 알 수 없어 GameConsoleInput에 가서 코드를 파악해야 하는데요.
이러한 점을 없애기 위해 어떤 인풋인지 정보가 드러나면 좋을 거 같습니다!

osandra replied:
회의에서 말한 것처럼 분리하는 방식으로 수정했습니다 ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 횟수를 "10"과 같은 숫자로 제대로 입력할 때까지 입력을 계속 받아, 횟수 값을 반환한다.
+    private fun getTimesInput(): Int {
```
상위 코드의 흐름에 맞게 getTimesInput()과 getCarNamesInput()의 위치를 바꾸는 건 어떤가요?

osandra replied:
위치를 바꾸는 게 더 나을 것 같네요! 수정했습니다 ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 입력받은 값인 횟수가 숫자 형태가 아니라면 null, 숫자라면 숫자를 반환한다.
+    fun getNumberOfTimes(input: String?): Int? {
```
상위 함수가 private 인데 여기는 public인 이유가 있을까요?
혹시 테스트를 위해서 열어둔 것이라면 상위 public fun을 테스트하는 방향으로 가보는 건 어떨까요?
view 쪽 코드는 테스트가 필수가 아니라(요구사항 한정) 테스트를 위해 public으로 변경할 필요는 없을 거 같아요ㅎㅎ

osandra replied:
getNumberOfTimes 를 테스트하기 위해서 열어두었습니다 ㅎㅎ 정확하게 알아보셨네요😂  
상위 함수에서 readLine()을 바로 사용해서 테스트하기 까다로워서 getNumberOfTimes 함수로 테스트를 진행했어요!
그럼 뷰 쪽 테스트는 제거하는 방향으로 진행해볼게요 ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+
+    // 이름이 5자 이상이거나 사용자가 값을 입력하지 않으면 빈 배열을 반환하고 정상적으로 입력했을 경우 리스트 형태로 반환한다.
+    fun getListOfNames(input: String?): List<String> {
+        val nameList = input?.split(",")
```
변수명에 컬렉션명을 넣는 것은 지양해야 한다고 합니다!
다만 저도 테스트 하나에 못 참고 넣었더라고요 나중에 보니까... 같이 지워나가봅시다ㅋㅋㅋ

osandra replied:
앗 그렇군요 ㅋㅋ 지워나가야겠네요! ␡

SukWooJung replied:
앗 무디의 코드 리뷰를 보고 댓글 달 듯이 남긴 리뷰인뎈ㅋㅋ 따로 달리네요 출처가 '클린코드' 맞나요 ??ㅎㅎ 저도 클린코드에서 컬렉션명 넣지 말라고 해서,, 억지로 복수형으로 변수명 선언한 것 같네욬ㅋㅋ

perenok commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+        return nameList.orEmpty()
+    }
+
+    // 이름이 5자 이상인 값이 입력되었다면 false 를 리턴하고, 정상적으로 입력되었다면 true 를 리턴합니다.
```
현재 메서드마다 주석이 달려있는데요. 주석은 다른 개발자들의 이해를 돕기도 하지만 그만큼 관리 포인트가 늘어나기도 합니다. 코드가 바뀌게 되면 주석도 바꿔야 하니까요. 코드와 다르게 주석은 내용이 맞지 않더라도 컴파일러가 잡아주지 못해 개발자들이 하나하나 체크해야 하니 메서드명에 의미를 최대한 담아보고 부족한 부분에만 주석을 달아도 좋을 것 같아요!

osandra replied:
주석이 관리 포인트가 될 수 있다는 부분은 생각하지 못했네요. 
코드만 봤을 때 이해하기 쉽지 않거나 부족한 부분에만 주석을 달고 메서드 명을 주의해서 작성하는 방향으로 진행해야겠네요 😄👍

perenok commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+}
+
+fun main() {
+    val consoleInput = GameConsoleInput()
+    val inputDto = consoleInput.getInputs()
```
```suggestion
    val inputDto = GameConsoleInput().getInputs()
```
다른 곳에서 사용되지 않는다면 인라인으로 한번에 선언해도 될 것 같아요ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+        printEachRound()
+    }
+
+    // 우승자 출력, 공동 우승이면 여러 명을 출력한다.
+    private fun printTotalResult() {
+        val maxMovedValue = cars.maxBy { it.movedValue }?.movedValue
+        val winnerNameList = cars.filter { it.movedValue == maxMovedValue }.map { it.name }.toList()
+        val winnerNamesOutput = winnerNameList.joinToString(separator = ", ")
+        println("${winnerNamesOutput}가 최종 우승했습니다.")
+    }
+
+    private fun printEachRound() {
+        val output = StringBuilder()
+        for (car in cars)
+            output.append("$car\n")
+        println("$output")
+    }
```
Input을 따로 UI 클래스로 빼둔 것 처럼 Output도 빼면 GameApplication 클래스가 본인의 역할에 보다 집중할 수 있을 것 같아요ㅎㅎ

osandra replied:
GameOutputView 를 생성하는 방식으로 수정했습니다 ㅎㅎ

perenok commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+    // 우승자 출력, 공동 우승이면 여러 명을 출력한다.
+    private fun printTotalResult() {
+        val maxMovedValue = cars.maxBy { it.movedValue }?.movedValue
+        val winnerNameList = cars.filter { it.movedValue == maxMovedValue }.map { it.name }.toList()
+        val winnerNamesOutput = winnerNameList.joinToString(separator = ", ")
+        println("${winnerNamesOutput}가 최종 우승했습니다.")
```
```suggestion
        val winnerNames = cars.filter { it.movedValue == maxMovedValue }
            .map { it.name }
            .toList()
            .joinToString(separator = ", ")
        println("${winnerNames}가 최종 우승했습니다.")
```
Stream을 쓸 때는 하나의 도트(.)마다 줄바꿈을 해서 가독성을 높였는데요. 코틀린에서도 동일하게 적용하는지 궁금하더라고요ㅋㅋㅋ 이번 피드백은 저런 방식이 코틀린에서 통하는지 정말 궁금해서 남긴거라 같이 얘기 나눠보면 좋을 거 같아요ㅋㅋㅋ

perenok commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+
+    private fun printEachRound() {
+        val output = StringBuilder()
+        for (car in cars)
+            output.append("$car\n")
```
```suggestion
        for (car in cars) {
            output.append("$car\n")
        }
```
한 줄 뿐이더라도 블럭을 선언하는 게 더 좋다고 합니다! 블럭이 없으면 다른 개발자가 실수할 여지가 있으니 한 줄 짜리 내용이더라도 블럭을 선언해보아요ㅎㅎ

SukWooJung replied:
무디의 꼼꼼한 리뷰 👍

perenok commented on src/main/kotlin/water/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,23 @@
+        private set
+
+    fun move(randomNumber: Int) {
+        if (randomNumber >= 4)
+            movedValue++
```
```suggestion
        if (randomNumber >= 4) {
            movedValue++
        }
```

perenok reviewed (commented):
안녕하세요 워터!
꼼꼼한 예외처리와 테스트에 감탄했습니다!!
제가 무심코 넘긴 케이스들에 대해서도 테스트를 다 하신 걸 보고 반성했습니다ㅠㅠ
몇 가지 의견 남겼는데 같이 의견을 나눠봤으면 좋겠습니다.
내일 이야기 나눠봐요~~

SukWooJung commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+        var carNameList: List<String> = ArrayList<String>()
+        while (carNameList.isEmpty()) {
+            carNameList = getListOfNames(readLine())
+        }
```
```suggestion  
        var carNameList: List<String> = ArrayList<String>()
        while (carNameList.isEmpty()) {
             println(CAR_NAME_INPUT_INFO_MESSAGE)
            carNameList = getListOfNames(readLine())
        }
```

SukWooJung replied:
println(CAR_NAME_INPUT_INFO_MESSAGE) 문장을 while문 안쪽에 넣으면, getListOfNames 함수의 똑같은 println(CAR_NAME_INPUT_INFO_MESSAGE) 문장을 지울 수 있어요 !! 
코드 중복도 지우고, 출력을 여러번 하는게 아닌지 확인해야 하는 비용도 줄일 수 있어서 살며시 건의해 봅니당 ㅎㅎ

osandra replied:
오 그렇네요 ㅎㅎ 훨씬 좋군요! 수정했습니다 ㅎㅎ

SukWooJung commented on src/main/kotlin/water/racingcar/ui/GameConsoleInput.kt:
```diff
@@ -0,0 +1,72 @@
+    private fun getTimesInput(): Int {
+        println(PLAY_TIMES_INPUT_INFO_MESSAGE)
+        var times = getNumberOfTimes(readLine())
+        while (times == null) {
```
저도 예외발생 부분에 있어서 고민을 많이했는데요. 
사용자가 잘못 입력한 경우 검증한 결과를 워터처럼 boolean형으로 전달하거나 null값을 전달하여 로직을 구현할 지 
예외처리를 해야할지 고민을 했습니다. 
제가 느끼기엔 현재 프로젝트엔 아니지만 null값이 허용되는 입력을 받는 경우가 분명히 있을 수 있다는 생각이 들어서요, 함수에서 검증 결과를 null 값을 전달 하기보다, 예외로 확실하게 인식하는게 낫다는 생각을 했어요 ! 그래야 개발자도 이것(전달된 검증결과)이 예외인지 아닌지 빠르게 인식하고 일률적으로 처리할 수 있을 것 같아서요. 

저도 제 수준의 상식으로 말씀드린거라 ! 회의때 얘기해보면 좋을 것 같아요

osandra replied:
IllegalArgumentException을 상속하는 InvalidInputRacingCarException 커스텀 클래스를 만들고, 
커스텀 클래스를 상속받아 각각에 맞는 예외(EmptyCarNamesInputException, InvalidCarNamesLengthInputException 등)를 던지는 방향으로 수정했습니다!

SukWooJung commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+import water.racingcar.util.RandomGenerator
+
+class GameApplication(input: InputDto) {
+    private val cars = input.carNames.map { Car(it) }.toList()
```
```suggestion
    private val cars = input.carNames.map { Car(it) }
```

SukWooJung replied:
map 자체가 list를 만들어 주는 것 같아서 .toList()는 생략하셔도 될 것 같습니다 !

osandra replied:
오 몰랐네요! 수정했습니다 ㅎㅎ

SukWooJung commented on src/main/kotlin/water/racingcar/GameApplication.kt:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+fun main() {
```
제가 워터의 프로젝트에서 느낀 한가지는 전체적인 로직의 흐름에 있어서 Input의 비중이 높게 설정되어 있지 않은가 생각이 들었습니다. (지극히 개인적인 의견) 입력받아야할 Input 값 2개를 먼저 받고 그 이후 로직이 진행되는데요. 이를 쪼개서 비중을 줄이고 순차적으로 로직을 진행했어도 되지 않았을까 생각이 듭니다. 예를들면, 실제 게임에서 제가 자동차 4대를 열심히 만들었는데 몇 판할지 고민하던 중간에 pc가 꺼져버려서 게임이 리부팅 되었다고 가정하면. 만들어놓은 자동차 4대의 정보는 게임 판수와 별개로 저장이 되어 있었어야되지만 로직이 그렇지 않아 다 손실된 경우가 있을 수도 있다는 생각이들어서요. 저도 잘 모르지만 조금 더 모듈단위로 쪼개는 고민을 함께 해보는 것도 좋을 것 같다는 생각이 들었습니다! [옳고 그름이 아닌 순전히 제 의견입니다 !! ㅠㅠ]

SukWooJung reviewed (approved):
안녕하세요 워터 ~! 꼼꼼하게 요구사항을 읽으시고 설계하신게 느껴지네요, 변수명이나 함수명도 일일이 신경쓰셔서 작성하신게 느껴지고요. 제가 놓친 테스트 케이스도 너무 작성을 잘 하신 것 같아서 보고 배울게 많네요 ㅎㅎ

고생하셨습니다 워터 ~! 저희 무디꺼 보고 많이 배워서 다시 수정하면서 폭풍성장 해보죠 !!!  😄

osandra committed:
refactor: 불필요한 주석과 코드 제거 및 변수명(컬렉션명) 변경

osandra committed:
refactor: add newLine at end of file & delete empty.txt

osandra committed:
refactor: InputDto 제거 후 파라미터로 주입받는 방식으로 변경 및 불필요한 테스트 코드 삭제

뷰(GameConsoleInput) 관련 테스트와 GameApplication의 구현 변경(repeat을 통해 라운드 횟수만큼 반복)으로 인해 불필요해진 테스트 삭제

osandra committed:
refactor: 콘솔 입력받는 클래스 object로 정의하는 방식으로 변경하고 콘솔 출력하는 클래스 구현
콘솔 입력받는 클래스 이름 또한 가독성 있게 변경함(GameConsoleInput->GameInputView)

osandra committed:
refactor: 한 줄 코드 블럭 추가 및 repeat 관련 리뷰 반영

osandra committed:
refactor: RandomGenerator 객체 선언 방식(object)으로 변경 및 랜덤 범위 시작/끝 값을 변수로 설정

osandra committed:
refactor: 유효하지 않은 값을 입력받을 때 예외를 던지는 방향으로 수정
